A36-xds-for-servers.md

```diff
@@ -205,9 +205,54 @@ must be used. The [most specific `filter_chain_match`][matching logic] of the
 repeated [`filter_chains`][Listener.filter_chains] should be found for the
 specific connection and if one is not found, the `default_filter_chain` must be
 used if it is provided. If there is no `default_filter_chain` and no chain
-matches, then the connection should be closed without any bytes being sent. The
-following is a snippet of [all current fields of
-FilterChainMatch][FilterChainMatch], and if they will be handled specially:
+matches, then the connection should be closed without any bytes being sent.
+
+If the most-specific matching logic might not produce a unique result, then the
+Listener must be NACKed. This case can only occur if there are duplicate
+matchers. However, when checking for duplicates the matchers need to be
+normalized.
+
+For normalization, each matcher should be replaced with the Cartesian product of
+its present fields, which avoids repeating fields and converts the matchers into
+disjunctive normal form. That is, the matcher:
+
+```
+destination_ports: 80, 8080
+prefix_ranges: 192.168.0.0/16, 10.0.0.0/8
+source_type: EXTERNAL
+```
+
+Should be treated as four matchers:
+
+```
+destination_ports: 80
+prefix_ranges: 192.168.0.0/16
+source_type: EXTERNAL
+
+destination_ports: 80
+prefix_ranges: 10.0.0.0/8
+source_type: EXTERNAL
+
+destination_ports: 8080
+prefix_ranges: 192.168.0.0/16
+source_type: EXTERNAL
+
+destination_ports: 8080
+prefix_ranges: 10.0.0.0/8
+source_type: EXTERNAL
+```
+
+CIDRs need mulitple normalizations. The `prefix_len` should be adjusted to the
+valid range of 0-32 (inclusive) for IPv4 and 0-128 (inclusive) for IPv6. An
+absent `prefix_len` should be considered equivalent to 32 for IPv4 and 128 for
```

Comment:
Envoy doesn't seem to be checking for an absent `prefix_len` https://github.com/envoyproxy/envoy/blob/5b875969d4571fc5bda77d6340d100142ed98797/source/common/network/cidr_range.cc#L116 which means that an absent `prefix_len` would be taken as 0

Reply:
Missing `prefix_len` should imply 0 I think instead of it being treated as a full IP address. Or even better is to reject the config

Reply:
Good catch; that is surprising. I've filed https://github.com/envoyproxy/envoy/issues/15579 to work through this. I fully expect we'll end up treating this as `0`.

Reply:
https://github.com/envoyproxy/envoy/issues/15579 is now resolved with a doc update and unit test. So I've swapped the language here to `0`.